src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/PackagePropertyPage/PackageIconValueProvider.cs

```diff
@@ -0,0 +1,101 @@
+        {
+            string filename = Path.GetFileName(filePath);
+            // Make a slash-only value into empty string, so it won't get prepended onto the path.
+            packagePath = ("\\".Equals(packagePath, StringComparison.Ordinal) || "/".Equals(packagePath, StringComparison.Ordinal))
+                ? string.Empty : (packagePath ?? string.Empty);
+            // The assumption is that PackagePath does not contain a path to a file; only a directory path.
+            return Path.Combine(packagePath, filename);
+        }
+
+        private async Task<IProjectItem?> GetExistingNoneItemAsync(string existingPackageIcon)
+        {
+            foreach (IProjectItem noneItem in await _sourceItemsProvider.GetItemsAsync(NoneItem))
+            {
+                string pack = await noneItem.Metadata.GetEvaluatedPropertyValueAsync(PackPropertyName);
+                string itemIconFilename = Path.GetFileName(noneItem.EvaluatedInclude);
+                string existingIconFilename = Path.GetFileName(existingPackageIcon);
+                // Instead of doing pure equality between itemPackageIcon and existingPackageIcon, a user may update the PackagePath
+                // of the item and forget to update the PackageIcon to reflect those changes, or vice versa. Instead, if the filename
+                // of this packed None item and the filename of the PackageIcon match, consider those to be related to one another.
+                if (TrueValue.Equals(pack, StringComparison.OrdinalIgnoreCase) && itemIconFilename.Equals(existingIconFilename, StringComparison.Ordinal))
+                {
+                    return noneItem;
+                }
+            }
+
+            return null;
+        }
+
+        public override async Task<string?> OnSetPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties, IReadOnlyDictionary<string, string>? dimensionalConditions = null)
+        {
+            string relativePath = PathHelper.MakeRelative(_unconfiguredProject, unevaluatedPropertyValue);
+            string existingPackageIcon = await defaultProperties.GetEvaluatedPropertyValueAsync(PackageIconPropertyName);
+            IProjectItem? existingItem = await GetExistingNoneItemAsync(existingPackageIcon);
+            string packagePath = string.Empty;
+            if (existingItem != null)
+            {
+                string itemPackagePath = await existingItem.Metadata.GetEvaluatedPropertyValueAsync(PackagePathPropertyName);
+                string itemPackageIcon = CreatePackageIcon(existingItem.EvaluatedInclude, itemPackagePath);
+                // The new filepath is the same as the current. No item changes are required.
+                if (relativePath.Equals(existingItem.EvaluatedInclude, StringComparison.Ordinal))
+                {
+                    return itemPackageIcon;
+                }
```

Comment:
Work can be avoided here in some cases:

```suggestion
                // The new filepath is the same as the current. No item changes are required.
                if (relativePath.Equals(existingItem.EvaluatedInclude, StringComparison.Ordinal))
                {
                    return CreatePackageIcon(existingItem.EvaluatedInclude, itemPackagePath);
                }
```

Reply:
This was also a product of refactoring. Originally, `itemPackageIcon` was used as part of the logic after this if-statement. Thus, it was used in 2 places. Now, it is not. Fixed.